bpo-43950: Add documentation for PEP-657 (#27047)
python/cpython: Doc/whatsnew/3.11.rst

New Features
============

.. _whatsnew310-pep657:

> Enhanced error locations in tracebacks

Reviewer: By the let's either make one of the existing or add a new example about multiple annotations 
```
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    lel3(x)
    ^^^^^^^
  File "test.py", line 12, in lel3
    return lel2(x) / 23
           ^^^^^^^
  File "test.py", line 9, in lel2
    return 25 + lel(x) + lel(x)
                ^^^^^^
  File "test.py", line 6, in lel
    return 1 + foo(a,b,c=x['z']['x']['y']['z']['y'], d=e)
                         ^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```
(maybe with better naming)

Author: Let me adopt that example with some different names.

Author: Changed to this example with different names.

--------------------------------------

When printing tracebacks, the interpreter will now point to the exact expression
that caused the error instead of just the line. For example:

> .. code-block:: pycon

Reviewer: ```suggestion
.. code-block:: python
```

Author: > Within Python highlighting mode, interactive sessions are recognized automatically and highlighted appropriately. Normal Python code is only highlighted if it is parseable

https://www.sphinx-doc.org/en/1.4.9/markup/code.html

Today I learned, thanks.

    Traceback (most recent call last):
      File "distance.py", line 11, in <module>
        print(manhattan_distance(p1, p2))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "distance.py", line 6, in manhattan_distance
        return abs(point_1.x - point_2.x) + abs(point_1.y - point_2.y)
                               ^^^^^^^^^
    AttributeError: 'NoneType' object has no attribute 'x'

Previous versions of the interpreter would point to just the line making it
ambiguous which object was ``None``. These enhanced errors can also be helpful
when dealing with deeply nested dictionary objects:

> .. code-block:: pycon

Reviewer: ```suggestion
.. code-block:: python
```

    Traceback (most recent call last):
      File "json_handler.py", line 21, in <module>
        json_blob['a']['b']['c']['d']['e']
        ^^^^^^^^^^^^^^^^^^^^^^^^
    TypeError: 'NoneType' object is not subscriptable

> This new column information for expressions can be retrieved using the

Author: Not sure if the `co_positions()` method and memory/disk usage should be pointed out here.

Reviewer: I don't think we should mention about that in the what's new, maybe just as a little warning in the `co_positions()`

Author: Moved to a note in `co_positions()`.

              ^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "distance.py", line 6, in manhattan_distance
        return abs(point_1.x - point_2.x) + abs(point_1.y - point_2.y)
                               ^^^^^^^^^
    AttributeError: 'NoneType' object has no attribute 'x'

Previous versions of the interpreter would point to just the line making it
ambiguous which object was ``None``. These enhanced errors can also be helpful
when dealing with deeply nested dictionary objects and multiple function calls:

.. code-block:: python

    Traceback (most recent call last):
      File "query.py", line 37, in <module>
        magic_arithmetic('foo')
        ^^^^^^^^^^^^^^^^^^^^^^^
      File "query.py", line 18, in magic_arithmetic
        return add_counts(x) / 25
               ^^^^^^^^^^^^^
      File "query.py", line 24, in add_counts
        return 25 + query_user(user1) + query_user(user2)
                    ^^^^^^^^^^^^^^^^^
      File "query.py", line 32, in query_user
        return 1 + query_count(db, response['a']['b']['c']['user'], retry=True)
                                   ^^^^^^^^^^^^^^^^^^^^^^^
    TypeError: 'NoneType' object is not subscriptable

> 

Reviewer: ```
See :pep:`657` for more details. (Contributed by Pablo Galindo, Batuhan Taskaya
and Ammar Askar in :issue:`43950`.)
```

Also add this to the end of this section.

New Features
============

> .. _whatsnew310-pep657:

Reviewer: ```suggestion
.. _whatsnew311-pep657:
```

New Features
============

.. _whatsnew311-pep657:

Enhanced error locations in tracebacks

> --------------------------------------

Reviewer 2: We need to mention here at the end of this section how to deactivate it. Basically a very similar note as:

```
         .. note::
            Storing positions may increase the disk usage of compiled Python
            files and memory usage of the interpreter. This feature can be
            disabled with the ``-X no_debug_ranges`` command line flag.
```

Author: Added as a `.. note::`. Can also change to a paragraph if that is better.

when dealing with deeply nested dictionary objects and multiple function calls,

.. code-block:: python

    Traceback (most recent call last):
      File "query.py", line 37, in <module>
        magic_arithmetic('foo')
        ^^^^^^^^^^^^^^^^^^^^^^^
      File "query.py", line 18, in magic_arithmetic
        return add_counts(x) / 25
               ^^^^^^^^^^^^^
      File "query.py", line 24, in add_counts
        return 25 + query_user(user1) + query_user(user2)
                    ^^^^^^^^^^^^^^^^^
      File "query.py", line 32, in query_user
        return 1 + query_count(db, response['a']['b']['c']['user'], retry=True)
                                   ~~~~~~~~~~~~~~~~~~^^^^^
    TypeError: 'NoneType' object is not subscriptable

as well as complex arithmetic expressions:

.. code-block:: python

    Traceback (most recent call last):
      File "calculation.py", line 54, in <module>
        result = (x / y / z) * (a / b / c)
                  ~~~~~~^~~
    ZeroDivisionError: division by zero

See :pep:`657` for more details. (Contributed by Pablo Galindo, Batuhan Taskaya
and Ammar Askar in :issue:`43950`.)

> Column information for code
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Column information for code objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

      File "query.py", line 18, in magic_arithmetic
        return add_counts(x) / 25
               ^^^^^^^^^^^^^
      File "query.py", line 24, in add_counts
        return 25 + query_user(user1) + query_user(user2)
                    ^^^^^^^^^^^^^^^^^
      File "query.py", line 32, in query_user
        return 1 + query_count(db, response['a']['b']['c']['user'], retry=True)
                                   ~~~~~~~~~~~~~~~~~~^^^^^
    TypeError: 'NoneType' object is not subscriptable

as well as complex arithmetic expressions:

.. code-block:: python

    Traceback (most recent call last):
      File "calculation.py", line 54, in <module>
        result = (x / y / z) * (a / b / c)
                  ~~~~~~^~~
    ZeroDivisionError: division by zero

See :pep:`657` for more details. (Contributed by Pablo Galindo, Batuhan Taskaya
and Ammar Askar in :issue:`43950`.)

.. note::
   This feature requires storing column positions in code objects which may
   result in a small increase of disk usage of compiled Python files or
   interpreter memory usage. To avoid storing the extra information and/or
   deactivate printing the extra traceback information, the

>    ``-X no_debug_ranges`` command line flag or the ``PYTHONNODEBUGRANGES``

Reviewer 2: ```suggestion
   ``-X no_debug_ranges`` command line flag or the :envvar:`PYTHONNODEBUGRANGES`
```

>    ``-X no_debug_ranges`` command line flag or the :envvar:`PYTHONNODEBUGRANGES`

Reviewer 2: ```suggestion
   :option:`-X` ``no_debug_ranges`` command line flag or the :envvar:`PYTHONNODEBUGRANGES`
```

   environment variable can be used.

Column information for code objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The information used by the enhanced traceback feature is made available as a
general API that can be used to correlate bytecode instructions with source
code. This information can be retrieved using:

- The :meth:`codeobject.co_positions` method in Python.
- The :c:func:`PyCode_Addr2Location` function in the C-API.

> The ``-X no_debug_ranges`` option and the environment variable

Reviewer 2: ```suggestion
The :option:`-X` ``no_debug_ranges`` option and the environment variable
```

Author: Nice, I was looking for the `:thing:` for this.

See :pep:`657` for more details. (Contributed by Pablo Galindo, Batuhan Taskaya
and Ammar Askar in :issue:`43950`.)

.. note::

>    This feature requires storing column positions in code objects which may
>    increase disk usage of compiled Python files or interpreter memory usage.
>    It can be disabled with the ``-X no_debug_ranges`` command line flag.

Reviewer 2: ```suggestion
   This feature requires storing column positions in code objects which may
   result in a small increase of disk usage of compiled Python files or interpreter memory usage.
   To avoid storing the extra information or to deactivate printing the extra traceback information,
   the feature can be fully disabled with the ``-X no_debug_ranges`` command line flag or the ``PYTHONNODEBUGRANGES`` environment variable.
```

Reviewer 2: Copy this also to the note in the `codeobject.co_positions()` docs

Author: Done.
